File: webui/src/features/board/types/style.ts
/**
 * @file Style types for the board feature.
 */

import { resolveFamilyShade } from "../lib/colors/tailwind"

/**
 * Type of node in the board.
 */
export type NodeType =
  | "rectangle"
  | "folder"
  | "text"
  | "sheet"
  | "slide"
  | "ellipse"
  | "diamond"
  | "soft-diamond"
  | "tag"
  | "layered-circle"
  | "image"
  | "icon"
  | "layered-rectangle"
  | "thought-cloud"
  | "capsule"
  | "layered-diamond"
  | "code-sandbox"
  | "widget"
  | "mini-app"
  | "ink"

/**
 * Stroke style for the node.
 */
export type StrokeStyle = "solid" | "dashed" | "dotted"

/**
 * Fill style for the node.
 */
export type FillStyle = "solid" | "hachure" | "cross-hatch" | "zigzag" | "dots"

/**
 * Text alignment options for the node.
 */
export type TextAlign = "left" | "center" | "right"

/**
 * Font size options for the node.
 */
export type FontSize = "S" | "M" | "L" | "XL"

// Convert font size to Tailwind CSS class
export function fontSizeToTwClass(size?: FontSize): string {
  switch (size) {
    case "S":
      return "text-sm"
    case "M":
      return "text-base"
    case "L":
      return "text-2xl"
    case "XL":
      return "text-4xl"
    default:
      return "text-base"
  }
}


/**
 * Font family options for the node.
 */
export type FontFamily = "handwriting" | "sans-serif" | "serif" | "monospace" | "informal"


// Convert font family to Tailwind CSS class
export function fontFamilyToTwClass(family?: FontFamily): string {
  switch (family) {
    case "handwriting":
      return "font-handwriting"
    case "sans-serif":
      return "font-sans"
    case "serif":
      return "font-serif"
    case "monospace":
      return "font-mono"
    case "informal":
      return "font-informal"
    default:
      return "font-sans"
  }
}


/**
 * Text style options for the node.
 */
export type TextStyle = "normal" | "bold" | "italic" | "underline" | "strikethrough"


// Convert text style to Tailwind CSS class
export function textStyleToTwClass(style?: TextStyle): string {
  switch (style) {
    case "normal":
      return "font-normal"
    case "bold":
      return "font-bold"
    case "italic":
      return "italic"
    case "underline":
      return "underline"
    case "strikethrough":
      return "line-through"
    default:
      return "font-normal"
  }
}


/**
 * Sloppiness (roughness) options for the node.
 */
export const SloppyPresets = [0, 0.5, 1]

export type Sloppiness = typeof SloppyPresets[number]

/**
 * Stroke width options for the node.
 */
export const StrokeWidthPresets = [1, 2, 4]

export type StrokeWidth = typeof StrokeWidthPresets[number]


export interface BaseStyle {
  angle: number
  strokeColor: string
  strokeWidth: number
  strokeStyle: StrokeStyle
  backgroundColor: string
  fillStyle: FillStyle
  roughness: number
  roundness: number
  opacity: number
  groupIds: string[]
  fontFamily: FontFamily
  fontSize: FontSize
  textAlign: TextAlign
  textColor: string
  textStyle: TextStyle
}


/**
 * Interface for the style of a node in the board.
 */
export interface Style extends BaseStyle {
  type: NodeType
}


// Arrowhead types for links
export type ArrowheadType = 'none' | 'arrow' | 'barb' | 'arrow-filled'


/**
 * Interface for the style of a link in the board.
 */
export interface LinkStyle extends BaseStyle {
  type: "arrow"
  sourceArrowhead: ArrowheadType
  targetArrowhead: ArrowheadType
  pathStyle: "bezier" | "straight" | "polyline"
}


export const TRANSPARENT_HEX = "#00000000"


const BLACK_HEX = "#000000"


const ORANGE_200 = resolveFamilyShade("orange", 200) ?? "#fed7aa"


const BLUE_200 = resolveFamilyShade("blue", 200) ?? "#bfdbfe"


const STONE_200 = resolveFamilyShade("stone", 200) ?? "#e7e5e4"


const STONE_800 = resolveFamilyShade("stone", 800) ?? "#292524"


const TEAL_200 = resolveFamilyShade("teal", 200) ?? "#99f6e4"


const SLATE_400 = resolveFamilyShade("slate", 400) ?? "#94a3b8"


const ROSE_PINE_LIGHT = "#faf4ed"


/**
 * Default style for nodes in the board.
 */
export const createDefaultStyle = ({
  type = "rectangle"
}: {
  type?: NodeType
}): Style => {
  const defaultOptions = {
    type: type,
    angle: 0.0,
    strokeColor: TRANSPARENT_HEX,
    strokeWidth: 2,
    strokeStyle: "solid",
    backgroundColor: ORANGE_200,
    fillStyle: "solid",
    textColor: BLACK_HEX,
    textStyle: "normal",
    opacity: 100,
    groupIds: []
  }

  switch (type) {
    case "rectangle":
    case "folder":
    case "layered-rectangle":
    case "layered-circle":
    case "tag":
      return {
        ...defaultOptions,
        roughness: 0.5,
        roundness: 0,
        fontFamily: "handwriting",
        fontSize: "M",
        textAlign: "center",
        backgroundColor: BLUE_200,
      } as Style
    case "thought-cloud":
      return {
        ...defaultOptions,
        roughness: 0.5,
        roundness: 0,
        fontFamily: "handwriting",
        fontSize: "M",
        textAlign: "center",
        backgroundColor: STONE_200
      } as Style
    case "capsule":
      return {
        ...defaultOptions,
        roughness: 0.5,
        roundness: 0,
        fontFamily: "handwriting",
        fontSize: "M",
        textAlign: "center",
        backgroundColor: TEAL_200
      } as Style
    case "ellipse":
      return {
        ...defaultOptions,
        roughness: 0.5,
        roundness: 0,
        fontFamily: "handwriting",
        fontSize: "M",
        textAlign: "center"
      } as Style
    case "diamond":
    case "layered-diamond":
    case "soft-diamond":
      return {
        ...defaultOptions,
        roughness: 0.5,
        roundness: 0,
        fontFamily: "handwriting",
        fontSize: "M",
        textAlign: "center"
      } as Style
    case "sheet":
      return {
        ...defaultOptions,
        roughness: 0,
        roundness: 0,
        fontFamily: "handwriting",
        fontSize: "M",
        textAlign: "left",
        backgroundColor: TRANSPARENT_HEX,
        strokeColor: TRANSPARENT_HEX,
      } as Style
    case "text":
    case "ink":
      return {
        ...defaultOptions,
        roughness: 0,
        roundness: 0,
        fontFamily: "handwriting",
        fontSize: "M",
        textAlign: "left",
        backgroundColor: TRANSPARENT_HEX,
        strokeColor: type === "ink" ? STONE_800 : TRANSPARENT_HEX,
      } as Style
    case "image":
      return {
        ...defaultOptions,
        roughness: 1,
        roundness: 0,
        fontFamily: "handwriting",
        fontSize: "M",
        textAlign: "center",
        backgroundColor: TRANSPARENT_HEX,
        strokeColor: TRANSPARENT_HEX,
      } as Style
    case "icon":
      return {
        ...defaultOptions,
        roughness: 1,
        roundness: 0,
        fontFamily: "handwriting",
        fontSize: "M",
        textAlign: "center",
        backgroundColor: TRANSPARENT_HEX,
        strokeColor: TRANSPARENT_HEX,
      } as Style
    case "slide":
      return {
        ...defaultOptions,
        roughness: 0,
        roundness: 2,
        fontFamily: "sans-serif",
        fontSize: "M",
        textAlign: "left",
        backgroundColor: TRANSPARENT_HEX,
        strokeColor: SLATE_400,
        strokeStyle: "dashed",
        opacity: 90,
      } as Style
    case "code-sandbox":
    case "widget":
    case "mini-app":
      return {
        ...defaultOptions,
        roughness: 0,
        roundness: 1,
        fontFamily: type === "code-sandbox" ? "monospace" : "sans-serif",
        fontSize: "S",
        textAlign: "left",
        backgroundColor: ROSE_PINE_LIGHT,
        strokeColor: TRANSPARENT_HEX,
        opacity: 100,
      } as Style
  }
}


/**
 * Default style for links in the board.
 */
export const createDefaultLinkStyle = (): LinkStyle => ({
  type: "arrow",
  angle: 0.0,
  strokeColor: STONE_800,
  strokeWidth: 2,
  strokeStyle: "solid",
  backgroundColor: TRANSPARENT_HEX,
  fillStyle: "solid",
  roughness: 1,
  roundness: 0,
  opacity: 100,
  groupIds: [],
  fontFamily: "handwriting",
  fontSize: "M",
  textAlign: "center",
  textColor: BLACK_HEX,
  textStyle: "normal",
  sourceArrowhead: "none",
  targetArrowhead: "arrow-filled",
  pathStyle: "bezier"
})

File: backend/topix/datatypes/note/style.py
"""Classes representing the style of a node in a graph."""

from enum import StrEnum
from typing import Literal

from pydantic import BaseModel

from topix.utils.colors import BLUE_200


class NodeType(StrEnum):
    """Enumeration for node types."""

    RECTANGLE = "rectangle"
    FOLDER = "folder"
    ELLIPSE = "ellipse"
    DIAMOND = "diamond"
    SHEET = "sheet"
    TEXT = "text"
    IMAGE = "image"
    ICON = "icon"
    LAYERED_RECTANGLE = "layered-rectangle"
    THOUGHT_CLOUD = "thought-cloud"
    CAPSULE = "capsule"
    LAYERED_DIAMOND = "layered-diamond"
    SOFT_DIAMOND = "soft-diamond"
    LAYERED_CIRCLE = "layered-circle"
    TAG = "tag"
    SLIDE = "slide"
    CODE_SANDBOX = "code-sandbox"
    WIDGET = "widget"
    MINI_APP = "mini-app"
    INK = "ink"


class StrokeStyle(StrEnum):
    """Enumeration for stroke styles."""

    SOLID = "solid"
    DASHED = "dashed"
    DOTTED = "dotted"


class FillStyle(StrEnum):
    """Enumeration for fill styles."""

    SOLID = "solid"
    HACHURE = "hachure"
    CROSSHATCH = "cross-hatch"
    ZIGZAG = "zigzag"
    DOTS = "dots"


class TextAlign(StrEnum):
    """Enumeration for text alignment."""

    LEFT = "left"
    CENTER = "center"
    RIGHT = "right"


class FontSize(StrEnum):
    """Enumeration for font sizes."""

    S = "S"
    M = "M"
    L = "L"
    XL = "XL"


class FontFamily(StrEnum):
    """Enumeration for font families."""

    HANDWRITING = "handwriting"
    SANS_SERIF = "sans-serif"
    SERIF = "serif"
    MONOSPACE = "monospace"
    INFORMAL = "informal"


class Style(BaseModel):
    """Style of a node in a graph.

    Attributes:
        type (str): Type of the node, default is "rectangle".
        angle (float): Rotation angle of the node in degrees.
        stroke_color (str): Color of the node's border.
        background_color (str): Background color of the node.
        fill_style (str): Fill style of the node, e.g., "solid"
        stroke_width (int): Width of the node's border.
        stroke_style (str): Style of the node's border, can be "solid" or "dashed".
        roughness (float): Roughness of the node, between 0 and 2.
        roundness (float): Roundness of the node's corners. 0 = sharp;
            1 = subtle (used by code-sandbox / widget); 3 = fully rounded
            (the user-facing "Rounded" toggle).
        opacity (int): Opacity of the node, default is 100%.
        group_ids (list[str]): List of group IDs this node belongs to.
        font_family (str): Font family used for text in the node.
        text_align (str): Text alignment within the node,
            can be "left", "center", or "right".
        color (str | None): Text color, if applicable.

    """

    # Defaults aligned with frontend createDefaultStyle() for rectangle nodes:
    # sharp corners (roundness 0) and the paper-adapted blue-200 fill.
    type: NodeType = NodeType.RECTANGLE
    angle: float = 0.0
    stroke_color: str = "#00000000"
    stroke_width: float = 2
    stroke_style: StrokeStyle = StrokeStyle.SOLID
    background_color: str = BLUE_200
    fill_style: FillStyle = FillStyle.SOLID
    roughness: float = 0.5
    roundness: float = 0.0
    opacity: float = 100
    group_ids: list[str] = []
    font_size: FontSize = FontSize.M
    font_family: FontFamily = FontFamily.HANDWRITING
    text_align: TextAlign = TextAlign.CENTER
    text_color: str = "#000000"
    text_style: str | None = "normal"


class Arrowhead(StrEnum):
    """Enumeration for arrowhead types."""

    ARROW = "arrow"
    BARB = "barb"
    ARROW_FILLED = "arrow-filled"
    NONE = "none"


class PathStyle(StrEnum):
    """Enumeration for path styles."""

    BEZIER = "bezier"
    STRAIGHT = "straight"
    POLYLINE = "polyline"


class LinkStyle(Style):
    """Style for links in a graph."""

    type: Literal["arrow"] = "arrow"

    # Defaults aligned with frontend createDefaultLinkStyle()
    stroke_color: str = "#292524"
    stroke_width: float = 2
    stroke_style: StrokeStyle = StrokeStyle.SOLID
    background_color: str = "#00000000"
    fill_style: FillStyle = FillStyle.SOLID
    roughness: float = 1
    roundness: float = 0
    opacity: float = 100
    group_ids: list[str] = []
    font_family: FontFamily = FontFamily.HANDWRITING
    font_size: FontSize = FontSize.M
    text_align: TextAlign = TextAlign.CENTER
    text_color: str = "#000000"
    text_style: str | None = "normal"

    source_arrowhead: Arrowhead = Arrowhead.NONE
    target_arrowhead: Arrowhead = Arrowhead.ARROW_FILLED

    path_style: PathStyle = PathStyle.BEZIER
